add unit test for autograd to make sure parent node can run ahead of the
reentrant node if reentrant node is on a different device

gqchen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35281 add unit test for autograd to make sure parent node can run ahead of the
reentrant node if reentrant node is on a different device**
reentrant node if reentrant node is on a different device**

reentrant node if reentrant node is on a different device

add unit test for autograd to make sure parent node can run ahead of the
reentrant node if reentrant node is on a different device

Differential Revision: [D20615939](https://our.internmc.facebook.com/intern/diff/D20615939/)

gqchen committed:
add unit test for autograd to make sure parent node can run ahead of the
reentrant node if reentrant node is on a different device

add unit test for autograd to make sure parent node can run ahead of the
reentrant node if reentrant node is on a different device

Differential Revision: [D20615939](https://our.internmc.facebook.com/intern/diff/D20615939/)

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
nit: In general I would advise doing `inp.clone()` to avoid any issues in case you do inplace later. But it's fine here if you want to leave it as-is.

albanD reviewed (commented):
This test as-is is interesting but it is slightly different from what I had in mind.
In particular, because the output of `ReentrantFn` is on the GPU, then this Node will run on the GPU thread.
So the worker thread that is blocked by the reentrant is the GPU one and not the CPU one.

Making `b0` a CPU Tensor would be interesting as well: the `ReentrantFn` would then run on the CPU worker thread and potentially prevent Parent from running on this same CPU thread.

wanchaol commented on test/test_autograd.py:
nit: non?

gqchen commented:
@albanD, can you clarify what you have in mind? There is an existing test "test_reentrant_priority" which covers the case where the reentrant is on CPU, and reentrant will be run before parent. 

My intention here is to make sure if reentrant is running on GPU, it should not block parent's CPU thread node. In our current implementation, it makes sure the ReadyQueue between parent and reentrant backwards are shared. This has not been covered by other tests.

gqchen committed:
Update on "add unit test for autograd to make sure parent node can run ahead of the
reentrant node if reentrant node is on a different device"

reentrant node if reentrant node is on a different device**

reentrant node if reentrant node is on a different device

add unit test for autograd to make sure parent node can run ahead of the
reentrant node if reentrant node is on a different device

Differential Revision: [D20615939](https://our.internmc.facebook.com/intern/diff/D20615939/)

[ghstack-poisoned]

gqchen commented:
This is part of the effort to add more unit tests to autograd. See #34367

albanD commented:
> can you clarify what you have in mind? There is an existing test "test_reentrant_priority" which covers the case where the reentrant is on CPU, and reentrant will be run before parent.

That would be a small extension to this one to make sure that if there is no more reentrant work to do (because it is on a different device for example), then the parent work is still executed.
